### DISCUSSION TITLE
Debug Channel doesn't work, just returns errors

### DISCUSSION THREAD

**Discussion by ppazos** [nextgenhealthcare/connect#5857] [General]
I have tried to Debug Channel on many channels and always getting exceptions. I'm not sure what should be done for the debugging to work.

My env is:

Mirth 4.3.0
JDK 11
Linux Mint

Log 1:

> [2023-07-24 10:50:24.432]  ERROR  (com.mirth.connect.server.transformers.JavaScriptFilterTransformer:119): Filter/Transformer error
> java.awt.HeadlessException
> 
> 	at java.desktop/java.awt.GraphicsEnvironment.checkHeadless(GraphicsEnvironment.java:208)
> 
> 	at java.desktop/java.awt.Window.<init>(Window.java:548)
> 
> 	at java.desktop/java.awt.Frame.<init>(Frame.java:423)
> 
> 	at java.desktop/javax.swing.JFrame.<init>(JFrame.java:224)
> 
> 	at org.mozilla.javascript.tools.debugger.SwingGui.<init>(SwingGui.java:210)
> 
> 	at org.mozilla.javascript.tools.debugger.Main.<init>(Main.java:41)
> 
> 	at org.mozilla.javascript.tools.debugger.MirthMain.<init>(MirthMain.java:15)
> 
> 	at org.mozilla.javascript.tools.debugger.MirthMain.createInstance(MirthMain.java:30)
> 
> 	at org.mozilla.javascript.tools.debugger.MirthMain.mirthMainEmbedded(MirthMain.java:55)
> 
> 	at org.mozilla.javascript.tools.debugger.MirthMain.mirthMainEmbedded(MirthMain.java:64)
> 
> 	at com.mirth.connect.server.util.javascript.JavaScriptUtil.getDebugger(JavaScriptUtil.java:856)
> 
> 	at com.mirth.connect.server.transformers.JavaScriptFilterTransformer.getDebugger(JavaScriptFilterTransformer.java:184)
> 
> 	at com.mirth.connect.server.transformers.JavaScriptFilterTransformer.initialize(JavaScriptFilterTransformer.java:103)
> 
> 	at com.mirth.connect.server.transformers.JavaScriptFilterTransformer.<init>(JavaScriptFilterTransformer.java:68)
> 
> 	at com.mirth.connect.server.controllers.DonkeyEngineController.createFilterTransformerExecutor(DonkeyEngineController.java:1462)
> 
> 	at com.mirth.connect.server.controllers.DonkeyEngineController.createSourceConnector(DonkeyEngineController.java:1401)
> 
> 	at com.mirth.connect.server.controllers.DonkeyEngineController.createChannelFromModel(DonkeyEngineController.java:1180)
> 
> 	at com.mirth.connect.server.controllers.DonkeyEngineController$DeployTask.doDeploy(DonkeyEngineController.java:1857)
> 
> 	at com.mirth.connect.server.controllers.DonkeyEngineController$DeployTask.execute(DonkeyEngineController.java:1835)
> 
> 	at com.mirth.connect.server.channel.ChannelTask.call(ChannelTask.java:67)
> 
> 	at com.mirth.connect.server.channel.ChannelTask.call(ChannelTask.java:16)
> 
> 	at java.base/java.util.concurrent.FutureTask.run(FutureTask.java:264)
> 
> 	at java.base/java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1128)
> 
> 	at java.base/java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:628)
> 
> 	at java.base/java.lang.Thread.run(Thread.java:834)

Log 2:

> [2023-07-24 10:50:24.437]  ERROR  (com.mirth.connect.server.channel.ErrorTaskHandler:26): com.mirth.connect.donkey.server.DeployException: Error initializing JavaScript Filter/Transformer
> 
> 	at com.mirth.connect.server.controllers.DonkeyEngineController$DeployTask.doDeploy(DonkeyEngineController.java:1860)
> 
> 	at com.mirth.connect.server.controllers.DonkeyEngineController$DeployTask.execute(DonkeyEngineController.java:1835)
> 
> 	at com.mirth.connect.server.channel.ChannelTask.call(ChannelTask.java:67)
> 
> 	at com.mirth.connect.server.channel.ChannelTask.call(ChannelTask.java:16)
> 
> 	at java.base/java.util.concurrent.FutureTask.run(FutureTask.java:264)
> 
> 	at java.base/java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1128)
> 
> 	at java.base/java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:628)
> 
> 	at java.base/java.lang.Thread.run(Thread.java:834)
> Caused by: com.mirth.connect.server.transformers.JavaScriptInitializationException: Error initializing JavaScript Filter/Transformer
> 
> 	at com.mirth.connect.server.transformers.JavaScriptFilterTransformer.initialize(JavaScriptFilterTransformer.java:120)
> 
> 	at com.mirth.connect.server.transformers.JavaScriptFilterTransformer.<init>(JavaScriptFilterTransformer.java:68)
> 
> 	at com.mirth.connect.server.controllers.DonkeyEngineController.createFilterTransformerExecutor(DonkeyEngineController.java:1462)
> 
> 	at com.mirth.connect.server.controllers.DonkeyEngineController.createSourceConnector(DonkeyEngineController.java:1401)
> 
> 	at com.mirth.connect.server.controllers.DonkeyEngineController.createChannelFromModel(DonkeyEngineController.java:1180)
> 
> 	at com.mirth.connect.server.controllers.DonkeyEngineController$DeployTask.doDeploy(DonkeyEngineController.java:1857)
> 
> 	... 7 more
> Caused by: java.awt.HeadlessException
> 
> 	at java.desktop/java.awt.GraphicsEnvironment.checkHeadless(GraphicsEnvironment.java:208)
> 
> 	at java.desktop/java.awt.Window.<init>(Window.java:548)
> 
> 	at java.desktop/java.awt.Frame.<init>(Frame.java:423)
> 
> 	at java.desktop/javax.swing.JFrame.<init>(JFrame.java:224)
> 
> 	at org.mozilla.javascript.tools.debugger.SwingGui.<init>(SwingGui.java:210)
> 
> 	at org.mozilla.javascript.tools.debugger.Main.<init>(Main.java:41)
> 
> 	at org.mozilla.javascript.tools.debugger.MirthMain.<init>(MirthMain.java:15)
> 
> 	at org.mozilla.javascript.tools.debugger.MirthMain.createInstance(MirthMain.java:30)
> 
> 	at org.mozilla.javascript.tools.debugger.MirthMain.mirthMainEmbedded(MirthMain.java:55)
> 
> 	at org.mozilla.javascript.tools.debugger.MirthMain.mirthMainEmbedded(MirthMain.java:64)
> 
> 	at com.mirth.connect.server.util.javascript.JavaScriptUtil.getDebugger(JavaScriptUtil.java:856)
> 
> 	at com.mirth.connect.server.transformers.JavaScriptFilterTransformer.getDebugger(JavaScriptFilterTransformer.java:184)
> 
> 	at com.mirth.connect.server.transformers.JavaScriptFilterTransformer.initialize(JavaScriptFilterTransformer.java:103)
> 
> 	... 12 more

**Comment by pacmano1** [comment]
Did you add the required JVM options that the debugger requires?   

https://docs.nextgen.com/bundle/Mirth_User_Guide_4_3_0/page/connect/connect/topics/t_to_edit_mcserver_file_java_connect.html

I know on windows you gotta start mcerver.exe (note not mcservice) at the command line, not via services (aka in Linux not via systemd).

And obviously the debuger must run on the same machine runnning the server process.

**Reply by ppazos** [reply]
Thanks @pacmano1 I'll try that config, though if you check the docs, it doesn't say that is a requirement for debugging https://docs.nextgen.com/bundle/Mirth_User_Guide_4_3_0/page/connect/connect/topics/c_mirth_connect_debugger_connect.html

"If your Java application does not interact directly with a user, use the following steps to change the Headless mode to false"

Not sure what that doc refers to as "your java application", I think they copy & pasted from some other site here, so the Mirth docs are confusing in this important area. It neither say having headless mode in true will lead to errors when trying to run the Channel Debug, or that this is actually required (there is a big IF at the beginning).

**Reply by pacmano1** [reply]
LOL, you are right, that's some poorly written documentation!  

See https://github.com/nextgenhealthcare/connect/wiki/4.0.0---What's-New#javascript-debugger
